File: IoT-Simulator/simulator.js
/**
 * Medico IoT Device Simulator
 * 
 * This script simulates an ESP32 device sending patient vitals to the backend.
 * Usage: node simulator.js
 * 
 * Requirements: Node.js installed
 */

const http = require('http');

// Configuration
const CONFIG = {
    backendUrl: 'http://localhost:8080',
    deviceId: 'ESP32_001',
    intervalMs: 10000, // Send data every 10 seconds
};

// Generate random vitals
function generateVitals(scenario = 'normal') {
    let heartRate, spo2, temperature;

    switch (scenario) {
        case 'critical':
            heartRate = Math.floor(Math.random() * 30) + 155; // 155-185
            spo2 = Math.floor(Math.random() * 5) + 85; // 85-89
            temperature = (Math.random() * 1.5 + 40).toFixed(1); // 40.0-41.5
            break;
        case 'moderate':
            heartRate = Math.floor(Math.random() * 30) + 125; // 125-155
            spo2 = Math.floor(Math.random() * 4) + 90; // 90-93
            temperature = (Math.random() * 1 + 38.5).toFixed(1); // 38.5-39.5
            break;
        case 'normal':
        default:
            heartRate = Math.floor(Math.random() * 30) + 60; // 60-90
            spo2 = Math.floor(Math.random() * 4) + 96; // 96-99
            temperature = (Math.random() * 0.8 + 36.2).toFixed(1); // 36.2-37.0
            break;
    }

    return {
        deviceId: CONFIG.deviceId,
        heartRate: heartRate,
        spo2: spo2,
        temperature: parseFloat(temperature)
    };
}

// Send vitals to backend
function sendVitals(vitals) {
    const data = JSON.stringify(vitals);

    const url = new URL(CONFIG.backendUrl + '/api/iot/vitals');

    const options = {
        hostname: url.hostname,
        port: url.port || 8080,
        path: url.pathname,
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(data),
            'X-API-Key': 'MEDICO_IOT_KEY_2026'
        }
    };

    const req = http.request(options, (res) => {
        let responseData = '';
        res.on('data', (chunk) => {
            responseData += chunk;
        });
        res.on('end', () => {
            const timestamp = new Date().toISOString();
            if (res.statusCode === 200) {
                console.log(`[${timestamp}] ✅ Vitals sent successfully`);
                console.log(`   HR: ${vitals.heartRate} BPM | SpO2: ${vitals.spo2}% | Temp: ${vitals.temperature}°C`);
                try {
                    const response = JSON.parse(responseData);
                    console.log(`   Risk Level: ${response.data?.riskLevel || 'Unknown'}`);
                } catch (e) { }
            } else {
                console.log(`[${timestamp}] ❌ Error: ${res.statusCode}`);
                console.log(`   Response: ${responseData}`);
            }
        });
    });

    req.on('error', (error) => {
        console.log(`[${new Date().toISOString()}] ❌ Connection error: ${error.message}`);
    });

    req.write(data);
    req.end();
}

// Parse command line arguments
const args = process.argv.slice(2);
let scenario = 'normal';
let continuous = true;

args.forEach(arg => {
    if (arg === '--critical') scenario = 'critical';
    if (arg === '--moderate') scenario = 'moderate';
    if (arg === '--normal') scenario = 'normal';
    if (arg === '--once') continuous = false;
    if (arg.startsWith('--device=')) CONFIG.deviceId = arg.split('=')[1];
    if (arg.startsWith('--url=')) CONFIG.backendUrl = arg.split('=')[1];
});

// Main
console.log('========================================');
console.log('    Medico IoT Device Simulator');
console.log('========================================');
console.log(`Device ID: ${CONFIG.deviceId}`);
console.log(`Backend URL: ${CONFIG.backendUrl}`);
console.log(`Scenario: ${scenario.toUpperCase()}`);
console.log(`Mode: ${continuous ? 'Continuous' : 'Single'}`);
console.log('----------------------------------------');

if (continuous) {
    console.log(`Sending vitals every ${CONFIG.intervalMs / 1000} seconds...`);
    console.log('Press Ctrl+C to stop\n');

    // Send immediately, then at intervals
    sendVitals(generateVitals(scenario));
    setInterval(() => {
        sendVitals(generateVitals(scenario));
    }, CONFIG.intervalMs);
} else {
    sendVitals(generateVitals(scenario));
}

File: IoT-Simulator/simulator.py
import requests
import random
import time

# ==============================
# CONFIGURATION
# ==============================

# Replace with your actual backend URL (localhost or ngrok)
# BACKEND_BASE_URL = "http://localhost:8080/api" 
BACKEND_BASE_URL = "https://f26f9576662d.ngrok-free.app/api"

API_KEY = "MEDICO_IOT_KEY_2026"
DEVICE_ID = "RASPBERRY_PI_001"

SEND_INTERVAL_SECONDS = 5  # seconds between sends

# Derived URLs
VITALS_URL = f"{BACKEND_BASE_URL}/iot/vitals"
STATUS_URL = f"{BACKEND_BASE_URL}/iot/device-status/{DEVICE_ID}"

# ==============================
# HELPER FUNCTIONS
# ==============================

def check_device_status():
    """
    Checks if the device is currently assigned to an active patient.
    Returns True if active, False otherwise.
    """
    try:
        headers = {
            "Content-Type": "application/json",
            "X-API-Key": API_KEY
        }
        response = requests.get(STATUS_URL, headers=headers, timeout=5)
        
        if response.status_code == 200:
            result = response.json()
            # Expected format: { "success": true, "data": true }
            is_active = result.get("data", False)
            return is_active
        else:
            print(f"⚠️ Status check failed: {response.status_code}")
            return False
    except Exception as e:
        print(f"⚠️ Status check error: {e}")
        return False

def send_vitals_to_backend(temperature, heart_rate, spo2):
    payload = {
        "deviceId": DEVICE_ID,
        "temperature": temperature,
        "heartRate": heart_rate,
        "spo2": spo2
    }

    headers = {
        "Content-Type": "application/json",
        "X-API-Key": API_KEY
    }

    try:
        response = requests.post(
            VITALS_URL,
            json=payload,
            headers=headers,
            timeout=5
        )

        if response.status_code == 200:
            data = response.json()
            print("✅ Sent successfully")
            print(f"   Temp: {temperature}°C | HR: {heart_rate} BPM | SpO₂: {spo2}%")
            risk = data.get('data', {}).get('riskLevel', 'N/A')
            print(f"   Risk Level: {risk}")
        else:
            print("❌ Backend error")
            print("Status:", response.status_code)
            print("Response:", response.text)

    except requests.exceptions.Timeout:
        print("❌ Timeout: Backend not responding")
    except requests.exceptions.ConnectionError:
        print("❌ Connection error: Check server / network")
    except Exception as e:
        print("❌ Unexpected error:", e)

# ==============================
# DUMMY DATA GENERATORS
# ==============================

def generate_random_vitals():
    temperature = round(random.uniform(36.0, 41.5), 1)
    heart_rate = random.randint(60, 180)
    spo2 = random.randint(85, 100)
    return temperature, heart_rate, spo2

# ==============================
# MAIN LOOP
# ==============================

def main():
    print("🚀 Dummy IoT Vitals Sender Started")
    print("📡 Backend URL:", BACKEND_BASE_URL)
    print("🆔 Device ID:", DEVICE_ID)
    print("-" * 50)

    while True:
        # 1. Check if we should send data
        is_active = check_device_status()

        if is_active:
             # 2. Generate and Send
            temperature, heart_rate, spo2 = generate_random_vitals()
            send_vitals_to_backend(temperature, heart_rate, spo2)
            print(f"⏳ Waiting {SEND_INTERVAL_SECONDS}s...")
        else:
            print("💤 Device Standby - Waiting for patient assignment...")
        
        time.sleep(SEND_INTERVAL_SECONDS)

# ==============================
# ENTRY POINT
# ==============================

if __name__ == "__main__":
    main()
